### DISCUSSION TITLE
Custom rendering with ReactNodeViewRenderer inside table cell

### DISCUSSION THREAD

**Discussion by seleckis** [ueberdosis/tiptap#5186] [General Questions] [answered]
Is it possible to create custom rendering for table cell content? There is a `@tiptap/extension-table-cell` extension which can be extended by `TableCell.extend` and with `ReactNodeViewRenderer` I can render whatever I want. But I need to leave table cell as is and intercept content rendering of a cell. Should I extend `Paragraph` extension in this case and check if its parent is a table cell, or something? What if I don't know what is rendering inside table cell and I just want to wrap everything with another element and leave table cell functionality as is.

**Comment by nperez0111** [comment]
I don't fully understand why you would not do as you said and do a `TableCell.extend`?

If you are saying that because ReactNodeViewRenderer renders a div you cannot do that? [NodeViewWrapper](https://tiptap.dev/docs/editor/guide/node-views/react#adding-a-content-editable) allows an `as` prop so you can do something like `as="td"`

**Reply by seleckis** [reply]
By documentation it should be used like this:
```ts
import TTTableCell from "@tiptap/extension-table-cell";

function Component() {
  return (
    <NodeViewWrapper as="td">
      <NodeViewContent />
    </NodeViewWrapper>
  );
}

export const TableCell = TTTableCell.extend({
   addNodeView() {
      return ReactNodeViewRenderer(Component);
   },
});
```
In this case `ReactNodeViewRenderer` ads extra div and produces wrong table structure:
```html
<table>
  <tr>
    <div class="react-renderer node-tableCell">
      <td>
        ....
```

**Reply by bdbch** [reply]
I created an issue out of this and will look into it.

**Reply by bdbch** [reply, marked as answer]
OK double verified this.

You should absolutely be able to pass in

```js
return ReactNodeViewRenderer(Component, {
  as: 'td',
})
```

which creates a ReactRenderer wrapper as a `td` element.

**Comment by stefanomarra** [comment]
Hi @seleckis,
were you able to find a solution and make the `TableCell.extend` work?
I'm struggling to make everything work as expected

**Reply by seleckis** [reply]
`TableCell.extend` works, but I don't need to use it for custom rendering anymore, I just add custom attributes and commands. Maybe you can try to use [contentDOMElementTag](https://tiptap.dev/docs/editor/extensions/custom-extensions/node-views/react#changing-the-default-content-tag-for-a-node-view) to make external tag as `td` and use `<NodeViewWrapper as="div">` for internal container.

**Reply by israel-arvizu** [reply]
I figured something out that worked for me, not sure if this is what you asking, but I used the `as` parameter in the `ReactNodeViewRenderer` to render the outer react renderer layer as a `td` so the table structure remains the same, and then passed a custom component in TableCellView to have a custom component I can modify, like so:

  ```
const TableCellNode = TableCell.extend<TableCellOptions>({
...
addNodeView() {
    return ReactNodeViewRenderer(TableCellView, {
      as: "td",
      attrs: ({ node }) => ({
        colSpan: node.attrs.colspan || 1,
        rowSpan: node.attrs.rowspan || 1,
        style: node.attrs.colwidth
          ? `width: ${node.attrs.colwidth[0]}px; position: relative;`
          : "position: relative;",
        "data-colwidth": node.attrs.colwidth?.join(","),
        "data-id": node.attrs.id,
      }),
    });
  },
...
```

Basic Custom Component that I can start modifying:
```
export const TableCellView = ({ node }: NodeViewProps) => {
    return (
        <NodeViewWrapper style={{ display: "contents" }}>
            <NodeViewContent />
        </NodeViewWrapper>
    );
};

```
